src/community.c

```diff
@@ -2111,6 +2111,261 @@ int igraph_le_community_to_membership(const igraph_matrix_t *merges,
+  /* Declaration of variables */
+  long int no_of_nodes, i, j, k, kv1;
+  igraph_adjlist_t al;
+  double max_density;
+  igraph_bool_t res, running;
+  igraph_vector_t node_order, density, label_counters, dominant_labels, nonzero_labels;
+  igraph_vector_int_t com_to_numvertices;
+  
+  /* Initialization of variables needed for initial checking */
+  no_of_nodes = igraph_vcount(graph);
+
+  /* Checking input values */
+  if (no_of_nodes < 2) {
+    IGRAPH_ERROR("Empty and single vertex graphs are not supported.", IGRAPH_EINVAL);
+  }
+  if ((long int) no_of_communities < 1) {
+    IGRAPH_ERROR("'no_of_communities' must be greater than 0.", IGRAPH_EINVAL);
+  }
+  if ((long int) no_of_communities > no_of_nodes) {
+    IGRAPH_ERROR("'no_of_communities' can not be greater than number of nodes in " 
+      "the graph.", IGRAPH_EINVAL);
+  }
+  igraph_is_simple(graph, &res);
+  if (!res) {
+    IGRAPH_ERROR("Only simple graphs are supported.", IGRAPH_EINVAL);
+  }
+  igraph_is_connected(graph, &res, IGRAPH_WEAK);
+  if (!res) {
+    IGRAPH_ERROR("Disconnected graphs are not supported.", IGRAPH_EINVAL);
+  }
+  if (igraph_is_directed(graph)) {
+    IGRAPH_WARNING("Edge directions are ignored.");
+  }
+
+  /* Internal variables initialization */
+  max_density = 1.0;
+  running = 1;
+
+  /* Resize membership vector (number of nodes) */
+  IGRAPH_CHECK(igraph_vector_resize(membership, no_of_nodes));
+
+  /* Initialize density and com_to_numvertices vectors */
+  IGRAPH_CHECK(igraph_vector_init(&density, (long int) no_of_communities));
+  IGRAPH_FINALLY(igraph_vector_destroy, &density);
+  IGRAPH_CHECK(igraph_vector_int_init(&com_to_numvertices, (long int) no_of_communities));
+  IGRAPH_FINALLY(igraph_vector_destroy, &com_to_numvertices);
+
+  /* Initialize node ordering vector */
+  IGRAPH_CHECK(igraph_vector_init_seq(&node_order, 0, no_of_nodes-1));
+  IGRAPH_FINALLY(igraph_vector_destroy, &node_order);
+
+  /* Initialize the membership vector with 0 values */
+  igraph_vector_null(membership);
+  /* Initialize densities to max_density */
+  igraph_vector_fill(&density, max_density);
+
+  RNG_BEGIN();
+
+  /* Initialize com_to_numvertices and initialize communities into membership vector */
+  IGRAPH_CHECK(igraph_vector_shuffle(&node_order));
+  for (i=0; i<no_of_communities; i++) {
+    /* Initialize membership at initial nodes for each community
+     * where 0 refers to have no label*/
+    VECTOR(*membership)[(long int)VECTOR(node_order)[i]] = i+1.0;
+    /* Initialize com_to_numvertices list: Number of vertices for each community */
+    VECTOR(com_to_numvertices)[i] = 1;
+  }
+
+  /* Create an adjacency list representation for efficiency. */
+  IGRAPH_CHECK(igraph_adjlist_init(graph, &al, IGRAPH_ALL));
+  IGRAPH_FINALLY(igraph_adjlist_destroy, &al);
+
+  /* Create storage space for counting distinct labels and dominant ones */
+  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, 0);
```

Comment:
I'm not sure about this. My initial idea on the initialization of `dominant_labels` was to create a vector of no size that will be able to growth at most to size `no_of_communities`.

What is the advantage of initializing it to a size that will be immediately cleared? Maybe you can help to clarify it.

Reply:
I have not looked at the details of your code, but I can answer the general question *"What is the advantage of initializing it to a size that will be immediately cleared?"*.

The size of an allocated memory block cannot usually be increased. Therefore, in order to increase the size of a vector, an entirely new memory block must be allocated and the contents of the old one must be copied over. Growing the size of a vector is slow, and it is best avoided. A typical way to do this is to pre-allocate as much storage as we expect the vector could ever need. This is why we have `igraph_vector_reserve` (and similar functions for all libraries that provide a vector datatype).

Immediately reserving as much storage for `dominant_labels` as it could ever need prevents any future reallocations.

As I said, I didn't look at the details of your code and I am not making claims about real-world performance impact.

Reply:
Thanks for clearing that up!
Then, I'm still not sure about what might be better between option A and B. If I understood correctly, both options might be kind of equivalent.

**Option A**
```suggestion
  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, 0);
  IGRAPH_CHECK(igraph_vector_reserve(&dominant_labels, (long int) no_of_communities));
  IGRAPH_VECTOR_INIT_FINALLY(&nonzero_labels, 0);
  IGRAPH_CHECK(igraph_vector_reserve(&nonzero_labels, (long int) no_of_communities));
  ...
  /* Clear dominant_labels and nonzero_labels vectors */
  igraph_vector_clear(&dominant_labels);
  igraph_vector_clear(&nonzero_labels);
```

**Option B**
```suggestion
  IGRAPH_VECTOR_INIT_FINALLY(&dominant_labels, (long int) no_of_communities));
  IGRAPH_VECTOR_INIT_FINALLY(&nonzero_labels, (long int) no_of_communities));
  ...
  /* Clear dominant_labels and nonzero_labels vectors */
  igraph_vector_clear(&dominant_labels);
  igraph_vector_clear(&nonzero_labels);
```

Reply:
I would suggest Option B for its simplicity.